src/main/java/com/celements/navigation/TreeNode.java

```diff
@@ -19,123 +19,95 @@
+      @Nullable Integer position, @NotNull PartNameGetter strategy) {
+    this(docRef, parentRef, position, null, strategy);
   }
 
-  public TreeNode(@NotNull DocumentReference docRef, DocumentReference parentRef,
-      Integer position) {
-    this(docRef, getWebUtilsService().getRefLocalSerializer().serialize(parentRef), position);
+  public TreeNode(@NotNull DocumentReference docRef, @Nullable EntityReference parentRef,
+      @Nullable Integer position) {
+    this(docRef, parentRef, position, (String) null);
   }
 
-  public TreeNode(@NotNull DocumentReference docRef, SpaceReference parentRef, String partName,
-      Integer position) {
-    this(docRef, "", position);
-    if (!StringUtils.isEmpty(partName)) {
-      setPartName(partName);
-    }
+  private TreeNode(@NotNull DocumentReference docRef, @Nullable EntityReference parentRef,
+      @Nullable Integer position, @Nullable String partName, @Nullable PartNameGetter strategy) {
+    Preconditions.checkNotNull(docRef, "Document reference for TreeNode may not be null.");
+    this.docRef = ModelUtils.cloneReference(docRef, DocumentReference.class);
+    this.parentRef = (parentRef != null) ? ModelUtils.cloneReference(parentRef) : null;
+    this.position = MoreObjects.firstNonNull(position, new Integer(0));
+    this.partName = (strategy == null) ? Strings.nullToEmpty(partName) : partName;
+    this.partNameStrategy = strategy;
   }
 
-  /**
-   * @return fullName
-   * @deprecated since 2.14.0 use getDocumentReference instead
-   */
-  @Deprecated
-  public String getFullName() {
-    return docRef.getLastSpaceReference().getName() + "." + docRef.getName();
+  public boolean isEmptyParentRef() {
+    return (parentRef == null);
   }
 
+  @NotNull
   public EntityReference getParentRef() {
-    if ("".equals(parent)) {
-      return docRef.getLastSpaceReference();
+    EntityReference ref;
+    if (isEmptyParentRef()) {
+      ref = getDocumentReference().getLastSpaceReference();
     } else {
-      return getWebUtilsService().resolveDocumentReference(parent,
-          (WikiReference) docRef.getLastSpaceReference().getParent());
+      ref = parentRef;
     }
+    return ModelUtils.cloneReference(ref);
   }
 
-  public String getParent() {
-    return parent;
-  }
-
-  void setParent(String parent) {
-    if (parent == null) {
-      parent = "";
-    }
-    this.parent = parent;
-  }
-
+  @NotNull
   public Integer getPosition() {
-    if (position == null) {
-      position = 0;
-    }
     return position;
   }
 
-  void setPosition(Integer position) {
-    this.position = position;
-  }
-
-  public String getPartName(XWikiContext context) {
-    if (partName == null) {
-      if (partNameGetStrategy != null) {
-        partName = partNameGetStrategy.getPartName(getFullName(), context);
-      } else {
-        partName = "";
+  @NotNull
+  public String getPartName() {
+    if ((partName == null) && (partNameStrategy != null)) {
```

Comment:
This condition always is false since parteName never is null. See the constructor statement:
`this.partName = (strategy == null) ? Strings.nullToEmpty(partName) : partName;`